languages/python/sqlalchemy-oso/sqlalchemy_oso/roles2.py

```diff
@@ -0,0 +1,477 @@
+                -- I hope there's a way to do this in sqlalchemy but if not it wouldn't
+                -- really be too hard to generate the sql
+                select
+                    {id_query}
+                    {type_query}
+                from resources
+                where type in {str(tuple(child_types))}
+            ) select * from resources
+        ), allow_permission as (
+            -- Find the permission
+            select
+                p.id
+            from permissions p
+            where p.resource_type = :resource_type and p.name = :action
+        ), permission_roles as (
+            -- find roles with the permission
+            select
+                rp.role
+            from role_permissions rp
+            join allow_permission ap
+            where rp.permission_id = ap.id
+        ), relevant_roles as (
+            -- recursively find all roles that have the permission or
+            -- imply a role that has the permission
+            with recursive relevant_roles (role) as (
+                select
+                    role
+                from permission_roles
+                union
+                select
+                    ri.from_role
+                from role_implications ri
+                join relevant_roles rr
+                on ri.to_role = rr.role
+            ) select * from relevant_roles
+        ), user_in_role as (
+            -- check if the user has any of those roles on any of the relevant resources
+            select
+                ur.resource_type,
+                ur.resource_id,
+                ur.role
+            from user_roles ur
+            join relevant_roles rr
+            on rr.role = ur.role
+            join resources r
+            on r.type = ur.resource_type and r.id = ur.resource_id
+            where ur.user_id = :user_id
+        ) select * from user_in_role;
+        """
+
+        self.configured = True
+
+    def _role_allows(self, user, action, resource):
+        # @TODO: Figure out where this session should really come from.
+        assert self.session
+
+        user_pk_name = inspect(user.__class__).primary_key[0].name
+        user_id = getattr(user, user_pk_name)
+
+        resource_pk_name = inspect(resource.__class__).primary_key[0].name
+        resource_id = getattr(resource, resource_pk_name)
+
+        results = self.session.execute(
+            self.sql_query,
+            {
+                "user_id": user.id,
+                "action": action,
+                "resource_id": resource_id,
+                "resource_type": resource.__class__.__name__,
+            },
+        )
+        role = results.first()
+        if role:
+            return True
+        else:
+            return False
+
+    def enable(self):
+        class Roles:
+            @staticmethod
+            def role_allows(user, action, resource):
+                if not self.configured:
```

Comment:
We probably shouldn't have this be lazily configured. (Maybe if we have a debug mode it's fine).

Otherwise the first request that performs authorization will create a bunch of data before executing

Reply:
Or we need the flag of whether it's configured to be persisted in the database so we can tell on startup.